lib/auth/auth_with_roles.go

```diff
@@ -2077,57 +2076,143 @@ func (a *ServerWithRoles) NewKeepAliver(ctx context.Context) (types.KeepAliver,
+func (a *ServerWithRoles) desiredAccessInfo(ctx context.Context, req *proto.UserCertsRequest, user types.User) (*services.AccessInfo, error) {
+	if req.Username != a.context.User.GetName() {
+		if req.UseRoleRequests || len(req.RoleRequests) > 0 {
+			err := trace.AccessDenied("User %v tried to issue a cert for %v and added role requests. This is not supported.", a.context.User.GetName(), req.Username)
+			log.WithError(err).Warn()
+			return nil, err
 		}
-
-		// Do not allow combining role impersonation and access requests
-		// Note: additional requested roles will fail validation; it may be
-		// possible to support this in the future if needed.
 		if len(req.AccessRequests) > 0 {
-			log.Warningf("User %v tried to issue a cert with both role and access requests. This is not supported.", a.context.User.GetName())
-			return nil, nil, trace.AccessDenied("access denied")
+			err := trace.AccessDenied("User %v tried to issue a cert for %v and added access requests. This is not supported.", a.context.User.GetName(), req.Username)
+			log.WithError(err).Warn()
+			return nil, err
+		}
+		return a.desiredAccessInfoForImpersonation(req, user)
+	}
+	if req.UseRoleRequests || len(req.RoleRequests) > 0 {
+		if len(req.AccessRequests) > 0 {
+			err := trace.AccessDenied("User %v tried to issue a cert with both role and access requests. This is not supported.", a.context.User.GetName())
+			log.WithError(err).Warn()
+			return nil, err
 		}
+		return a.desiredAccessInfoForRoleRequest(req)
+	}
+	return a.desiredAccessInfoForUser(ctx, req, user)
+}
+
+// desiredAccessInfoForImpersonation returns the desired AccessInfo for an
+// impersonation request.
+func (a *ServerWithRoles) desiredAccessInfoForImpersonation(req *proto.UserCertsRequest, user types.User) (*services.AccessInfo, error) {
+	return &services.AccessInfo{
+		Roles:  user.GetRoles(),
+		Traits: user.GetTraits(),
+	}, nil
+}
+
+// desiredAccessInfoForRoleRequest returns the desired roles for a role request.
+func (a *ServerWithRoles) desiredAccessInfoForRoleRequest(req *proto.UserCertsRequest) (*services.AccessInfo, error) {
+	// If UseRoleRequests is set, make sure we don't return unusable certs: an
+	// identity without roles can't be parsed.
+	// DEPRECATED: consider making role requests without UseRoleRequests set an
+	// error in V11.
+	if req.UseRoleRequests && len(req.RoleRequests) == 0 {
+		return nil, trace.BadParameter("at least one role request is required")
+	}
+
+	// If role requests are provided, attempt to satisfy them instead of
+	// pulling them directly from the logged in identity. Role requests
+	// are intended to reduce allowed permissions so we'll accept them
+	// as-is for now (and ensure the user is allowed to assume them
+	// later).
+	// Note: traits are not currently set for role impersonation.
+	return &services.AccessInfo{
+		Roles: req.RoleRequests,
+	}, nil
+}
+
+// desiredRolesAndTraitsForUser returns the desired roles and traits for a user
```

Comment:
Looks like in refactoring, the function name referenced in this comment didn't get updated.

Reply:
thanks, fixed